### ISSUE TITLE
some question,pls

### ISSUE THREAD

**Issue by boboChina** [mysql-time-machine/replicator#2]
1、How to deal with all production quantity data initialization?
2、According to binlog position flexible setup synchronization?
3、https://mysql-time-machine.github.io/user-guide/replicator.This article does not mention how distributed deployment.
4、How to monitor the binlog synchronization delay, send alarm information at the same time

**Comment by bdevetak** [comment]
Hi,

1、How to deal with all production quantity data initialization?

You need to flush the database to the binlog. That way the initial copy of database is in the binlog before you start the replication. User guide at: https://mysql-time-machine.github.io/user-guide/binlog-flusher/

2、According to binlog position flexible setup synchronization?

Not sure I understand what you mean by "flexible setup synchronization"?

3、https://mysql-time-machine.github.io/user-guide/replicator.This article does not mention how distributed deployment.

Not sure what do you mean by distributed deployment? Are you referring to running HBase in distributed mode, or something else?

4、How to monitor the binlog synchronization delay, send alarm information at the same time

Replicator sends metrics to graphite. Two of them are most useful for this: 'replicatorReplicationDelay' and 'rowOpsCommittedToHbase'. Based on them you can create Bosun alerts.

**Comment by boboChina** [comment]
about 2：master.journal.name,master.position    .can refer to alibaba's canal project :   https://github.com/alibaba/canal

about 3: running mysql-time-machine in distributed mode.

about 4: get it

thx.

**Comment by bdevetak** [comment]
Q2. I looked at Canal and these refer to binlog file name and position. MySQL time machine stores them in zookeeper, similar as CanalMetaManager. So, if you don't provide binlog file name on startup it will load it and the position from the meta store. There are two meta store option: file (useful for debugging, or if replicator instances are on one machine) or zookeeper (for distributed deployment). The binlog file name and position that is stored in meta store is the last confirmed position (AFAIK the CanalMetaManager works the same way). So, if one instance dies, the stand-by will take leadership and continue for last confirmed position.

Confirm position has different options. If pseudo GTID is used (supported for HBase replication, will be added for Kafka in the future), then every 5s we have pseudo GTID emitted and is zookeeper, the last successfully committed pseudo GTID is stored. If pseudo GTID is not used (Kafka) then the binlog filename and position is stored in zookeeper. The frequency of storage is on binlog rotate event and on DDL event.

For pseudoGTID setup see mysql-scripts repo.

At the moment native MySQL GTID are not supported, only pseudo GTIDs. In the future support for native GTIDs is planned.

Created issue #4 to document pseudoGTIDs and faileover.


Q3. When replicator instance is started it tries to acquire a lock in zookeeper. If multiple instances are started, one that acquires a lock starts to replicate data from the last confirmed position. The other instances are waiting on stand by. If the leader instance dies, then one of the stand by instances acquires the lock and starts to replicate from the confirmed position. Confirmed position is the last binlog position which has been confirmed to be committed. That position is stored in zookeeper. The easiest way to setup this is by running replicator in docker and using some container orchestration like Marathon. In that case Marathon will restart the instance that died so your container pool always has the number of instances you have defined.

I created an issue #3 to document this setup at the documentation site.
